storage/libradb/src/ledger_store/ledger_info_test.rs

```diff
@@ -4,83 +4,129 @@
 use super::*;
 use crate::{change_set::ChangeSet, LibraDB};
 use libra_tools::tempdir::TempPath;
-use libra_types::proptest_types::{AccountInfoUniverse, LedgerInfoWithSignaturesGen};
+use libra_types::{
+    proptest_types::{AccountInfoUniverse, LedgerInfoWithSignaturesGen},
+    transaction::Version,
+};
 use proptest::{collection::vec, prelude::*};
+use std::path::Path;
 
 fn arb_ledger_infos_with_sigs() -> impl Strategy<Value = Vec<LedgerInfoWithSignatures>> {
     (
         any_with::<AccountInfoUniverse>(3),
         vec((any::<LedgerInfoWithSignaturesGen>(), 1..10usize), 1..100),
     )
         .prop_map(|(mut universe, gens)| {
-            gens.into_iter()
+            let ledger_infos_with_sigs: Vec<_> = gens
+                .into_iter()
                 .map(|(ledger_info_gen, block_size)| {
                     ledger_info_gen.materialize(&mut universe, block_size)
                 })
-                .collect()
+                .collect();
+            assert_eq!(get_first_epoch(&ledger_infos_with_sigs), 0);
+            ledger_infos_with_sigs
         })
 }
 
+fn get_first_epoch(ledger_infos_with_sigs: &[LedgerInfoWithSignatures]) -> u64 {
+    ledger_infos_with_sigs
+        .first()
+        .unwrap()
+        .ledger_info()
+        .epoch()
+}
+
+fn get_last_epoch(ledger_infos_with_sigs: &[LedgerInfoWithSignatures]) -> u64 {
+    ledger_infos_with_sigs.last().unwrap().ledger_info().epoch()
+}
+
+fn get_last_version(ledger_infos_with_sigs: &[LedgerInfoWithSignatures]) -> Version {
+    ledger_infos_with_sigs
+        .last()
+        .unwrap()
+        .ledger_info()
+        .version()
+}
+
 proptest! {
-    #![proptest_config(ProptestConfig::with_cases(10))]
+    #![proptest_config(ProptestConfig::with_cases(20))]
 
     #[test]
-    fn test_ledger_info_put_get_verify(
-        ledger_infos_with_sigs in arb_ledger_infos_with_sigs()
+    fn test_get_epoch_change_ledger_infos(
+        (ledger_infos_with_sigs, start_epoch, end_epoch) in arb_ledger_infos_with_sigs()
+            .prop_flat_map(|ledger_infos_with_sigs| {
+                let first_epoch = get_first_epoch(&ledger_infos_with_sigs);
+                let last_epoch = get_last_epoch(&ledger_infos_with_sigs);
+                (
+                    Just(ledger_infos_with_sigs),
+                    first_epoch..=last_epoch,
+                )
+            })
+            .prop_flat_map(|(ledger_infos_with_sigs, start_epoch)| {
```

Comment:
it took me a while to understand the flag_map stuff, probably worth a comment what're generated

Reply:
Yeah, actually it'd be helpful to look at https://altsysrq.github.io/proptest-book/proptest/tutorial/higher-order.html